clar2wasm/src/wasm_generator.rs

```diff
@@ -231,7 +231,28 @@ impl WasmGenerator {
                 },
                 args,
             )) => {
-                if let Some(word) = words::lookup(function_name) {
+                if let Some(simpleword) = words::lookup_simple(function_name) {
```

Comment:
In the classical terminology for tree traversal, the "simple" functions should use a simple, generic post-order "traverse" method, then define their own "visit" method, while the "complex" functions should define their own "traverse" method. What do you think about using these terms?

We could still have one common trait, `Word`, which defines the traverse method, then the trait `SimpleWord` implements `Word` with a generic post-order traversal, then calls a `visit_expr` method.

Reply:
I started using one trait for all words, however i could not find a way to cleanly express or "downcast" the words based on the traits, so i would say we keep it two different traits.

If so, what would you suggest the naming to be?

Reply:
I was thinking just `Word` and `SimpleWord`.

```rust
pub trait Word: Sync + core::fmt::Debug {
    fn name(&self) -> ClarityName;

    fn traverse(
        &self,
        generator: &mut WasmGenerator,
        builder: &mut InstrSeqBuilder,
        _expr: &SymbolicExpression,
        args: &[SymbolicExpression],
    ) -> Result<(), GeneratorError>;
}

pub trait SimpleWord: Word {
    fn traverse(
        &self,
        generator: &mut WasmGenerator,
        builder: &mut InstrSeqBuilder,
        expr: &SymbolicExpression,
        args: &[SymbolicExpression],
    ) -> Result<(), GeneratorError> {
        // traverse arguments
        for arg in args {
            generator.traverse_expr(builder, arg)?;
        }
        let arg_types: Result<Vec<TypeSignature>, GeneratorError> = args
            .iter()
            .map(|e| {
                generator.get_expr_type(e)
                    .cloned()
                    .ok_or(GeneratorError::InternalError(
                        "expected valid argument type".into(),
                    ))
            })
            .collect();
        let return_type = generator
            .get_expr_type(expr)
            .expect("Simple words must be typed")
            .clone();

        self.visit(generator, builder, &arg_types?, &return_type)
    }

    fn visit(
        &self,
        generator: &mut WasmGenerator,
        builder: &mut InstrSeqBuilder,
        arg_types: &[TypeSignature],
        return_type: &TypeSignature,
    ) -> Result<(), GeneratorError>;
}
```

Reply:
A default trait impl is still overridable, but would not work with mapping. I don't think that is a very good design

Reply:
Ok, I will finish reading through the rest of the changes to get the entire picture.

Reply:
What about making it a wrapper struct instead, like this? Can you explain "would not work with mapping"?

```rust
pub trait SimpleWord: Sync + core::fmt::Debug {
    fn name(&self) -> ClarityName;

    fn visit(
        &self,
        generator: &mut WasmGenerator,
        builder: &mut InstrSeqBuilder,
        arg_types: &[TypeSignature],
        return_type: &TypeSignature,
    ) -> Result<(), GeneratorError>;
}

pub struct SimpleWrapper<T: SimpleWord> {
    inner: T,
}

impl<T: SimpleWord> core::fmt::Debug for SimpleWrapper<T> {
    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
        self.inner.fmt(f)
    }
}

impl<T: SimpleWord> Word for SimpleWrapper<T> {
    fn name(&self) -> ClarityName {
        self.inner.name()
    }

    fn traverse(
        &self,
        generator: &mut WasmGenerator,
        builder: &mut InstrSeqBuilder,
        expr: &SymbolicExpression,
        args: &[SymbolicExpression],
    ) -> Result<(), GeneratorError> {
        // traverse arguments
        for arg in args {
            generator.traverse_expr(builder, arg)?;
        }
        let arg_types: Result<Vec<TypeSignature>, GeneratorError> = args
            .iter()
            .map(|e| {
                generator
                    .get_expr_type(e)
                    .cloned()
                    .ok_or(GeneratorError::InternalError(
                        "expected valid argument type".into(),
                    ))
            })
            .collect();
        let return_type = generator
            .get_expr_type(expr)
            .expect("Simple words must be typed")
            .clone();

        self.inner
            .visit(generator, builder, &arg_types?, &return_type)
    }
}
```

Reply:
We can revisit this later.